ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1009,6 +1018,20 @@ SnapshotManagementRequestHandler getSnapshotRequestHandler() {
     return snapshotRequestHandler;
   }
 
+  CompletableFuture<RaftClientReply> pauseLeaderElectionAsync(PauseLeaderElectionRequest request) throws IOException {
+    LOG.info("{} receive pauseLeaderElection {}", getMemberId(), request);
+    assertLifeCycleState(LifeCycle.States.RUNNING);
+    assertGroup(request.getRequestorId(), request.getRaftGroupId());
+
+    boolean pause = getRole().getLeaderElectionPauseState();
+    if (pause == request.getPause()) {
+      return CompletableFuture.completedFuture(newSuccessReply(request, request.getPausedServer()));
+    }
+    synchronized (this) {
+      return pauseLeaderElection.pauseLeaderElectionAsync(request);
+    }
+  }
```

Comment:
Let's rename this method to setLeaderElectionAsync.  The implementation could be very simple.
```
  CompletableFuture<RaftClientReply> setLeaderElectionAsync(LeaderElectionRequest request) throws IOException {
    LOG.info("{} receive setLeaderElection {}", getMemberId(), request);
    assertLifeCycleState(LifeCycle.States.RUNNING);
    assertGroup(request.getRequestorId(), request.getRaftGroupId());

    synchronized (this) {
      setLeaderElectionPause(request.getPause());
    }
    return CompletableFuture.completedFuture(newSuccessReply(request));
  }
```

Reply:
We actually could call  getRole().setLeaderElectionPause(request.getPause()) directly without synchronized.
```
  CompletableFuture<RaftClientReply> setLeaderElectionAsync(LeaderElectionRequest request) throws IOException {
    LOG.info("{} receive setLeaderElection {}", getMemberId(), request);
    assertLifeCycleState(LifeCycle.States.RUNNING);
    assertGroup(request.getRequestorId(), request.getRaftGroupId());

    getRole().setLeaderElectionPause(request.getPause());
    return CompletableFuture.completedFuture(newSuccessReply(request));
  }
```